portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryLocalServiceImpl.java

```diff
@@ -2771,6 +2778,87 @@ private long _getValidFileEntryTypeId(
 		}
 	}
 
+	private void _inheritRolesPermissions(
+			long inheritableParentFolderId, DLFileEntry dlFileEntry)
+		throws PortalException {
+
+		long companyId = dlFileEntry.getCompanyId();
+		String parentClassName = DLFolderConstants.getClassName();
+
+		Set<String> dlFolderModelActionIds = SetUtil.fromCollection(
+			ResourceActionsUtil.getModelResourceActions(
+				DLFolderConstants.getClassName()));
+
+		Set<String> dlFileEntryModelActionIds = SetUtil.fromCollection(
+			ResourceActionsUtil.getModelResourceActions(
+				DLFileEntryConstants.getClassName()));
+
+		if (inheritableParentFolderId ==
+				DLFolderConstants.DEFAULT_PARENT_FOLDER_ID) {
+
+			int count =
+				_resourcePermissionLocalService.getResourcePermissionsCount(
+					companyId, DLConstants.RESOURCE_NAME,
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(dlFileEntry.getGroupId()));
+
+			if (count == 0) {
+				_resourceLocalService.addResources(
+					companyId, dlFileEntry.getGroupId(), 0,
+					DLConstants.RESOURCE_NAME,
+					String.valueOf(dlFileEntry.getGroupId()), false, true,
+					true);
+			}
+
+			parentClassName = DLConstants.RESOURCE_NAME;
+			inheritableParentFolderId = dlFileEntry.getGroupId();
+		}
+
+		Map<Long, Set<String>> dlFolderRoleIdsToActionIds =
+			_resourcePermissionLocalService.
+				getAvailableResourcePermissionActionIds(
+					companyId, parentClassName,
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(inheritableParentFolderId),
+					dlFolderModelActionIds);
+
+		Map<Long, Set<String>> dlFileEntryRoleIdsToActionIds =
+			_resourcePermissionLocalService.
+				getAvailableResourcePermissionActionIds(
+					companyId, DLFileEntryConstants.getClassName(),
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(dlFileEntry.getFileEntryId()),
+					dlFileEntryModelActionIds);
+
+		Set<Long> dlFolderRoleIds = dlFolderRoleIdsToActionIds.keySet();
+
+		for (Long dlFolderRoleId : dlFolderRoleIds) {
+			Set<String> dlFolderActionIds = dlFolderRoleIdsToActionIds.get(
+				dlFolderRoleId);
+
+			dlFolderActionIds.retainAll(_commonPermissions);
+
+			Set<String> dlFileEntryActionIds = new HashSet<>(dlFolderActionIds);
+
+			if (dlFileEntryRoleIdsToActionIds.containsKey(dlFolderRoleId)) {
+				Set<String> dlFileEntryRoleIdToActionIds =
+					dlFileEntryRoleIdsToActionIds.get(dlFolderRoleId);
+
+				for (String actionId : dlFileEntryRoleIdToActionIds) {
+					if (!_commonPermissions.contains(actionId)) {
+						dlFileEntryActionIds.add(actionId);
+					}
+				}
```

Comment:
I think it should not. Because if we removeAll commonPermission in dlFileEntryRoleIdToActionIds. So then we have to add that new set to dlFileEntryActionIds again.

Reply:
I think `removeAll` is more explicit than the loop, and they make it clear that the goal is to remove the common permissions from the dlFileEntryRoleIdToActionIds set and add the remaining permissions to the dlFileEntryActionIds set.
In addition, the removeAll has a lower time complexity.

Reply:
Fixed it. Thanks